### PR TITLE
Test for determining effective io class

### PR DESCRIPTION
**Pull request by Deixx**
Requires https://github.com/Open-CAS/test-framework/pull/50

Signed-off-by: Daniel Madej <daniel.madej@intel.com>

### CODE REVIEW

**Comment by micrakow** [comment]
@CAS-Linux-Jenkins retest this please

File: test/functional/tests/io_class/test_io_classification.py [hunk 1]
```diff
@@ -569,3 +571,138 @@ def test_ioclass_conditions_and(filesystem):
...
+    with TestRun.LOGGER.step(f"Preparing {filesystem.name} filesystem "
+                             f"and mounting {core.system_path} at {mountpoint}"):
+        core.create_filesystem(filesystem)
+        core.mount(mountpoint)
+        fs_utils.create_directory(test_dir)
+        sync()
+
+    id_permutations = list(permutations(range(1, 5)))
+    i = 1
+    for permutation in id_permutations:
```
Comment (robertbaldyga) [inline test/functional/tests/io_class/test_io_classification.py]:
```python
for i, permutation in enumerate(permutations(range(1, 5)), start=1):
```

File: test/functional/tests/io_class/test_io_classification.py [hunk 2]
```diff
@@ -569,3 +571,138 @@ def test_ioclass_conditions_and(filesystem):
...
+            allocation=True,
+            rule=rules[2],
+            ioclass_config_path=ioclass_config_path,
+        )
+        ioclass_config.add_ioclass(
+            ioclass_id=permutation[3],
+            eviction_priority=22,
+            allocation=True,
+            rule=rules[3],
+            ioclass_config_path=ioclass_config_path,
```
Comment (robertbaldyga) [inline test/functional/tests/io_class/test_io_classification.py]: This looks like good candidate for `for` loop.

File: test/functional/tests/io_class/test_io_classification.py [hunk 3]
```diff
@@ -569,3 +571,138 @@ def test_ioclass_conditions_and(filesystem):
...
+
+
+@pytest.mark.require_disk("cache", DiskTypeSet([DiskType.optane, DiskType.nand]))
+@pytest.mark.require_disk("core", DiskTypeLowerThan("cache"))
+@pytest.mark.parametrize("filesystem", Filesystem)
+def test_ioclass_effective_ioclass(filesystem):
+    """
+    title: Effective IO class with multiple non-exclusive conditions
+    description: Test CAS ability to properly classify IO fulfilling multiple conditions based on
+    IO class ids and presence of '&done' annotation in IO class rules
```
Comment (robertbaldyga) [inline test/functional/tests/io_class/test_io_classification.py]: Please use `|` operator.

File: test/functional/tests/io_class/test_io_classification.py (hunk 1 above)
Comment (robertbaldyga) [reply test/functional/tests/io_class/test_io_classification.py]: Now you can do:
```python
for i, permutation in TestRun.iteration(enumerate(permutations(range(1, 5)), start=1)):
```

**Comment by katlapinka** [comment]
@CAS-Linux-Jenkins retest this please

File: test/functional/tests/io_class/test_io_classification.py [hunk 4]
```diff
@@ -569,3 +571,111 @@ def test_ioclass_conditions_and(filesystem):
         assert new_occupancy == base_occupancy, \
```
[the remark this answers is not in the document]
Comment (Deixx) [reply test/functional/tests/io_class/test_io_classification.py]: Done

File: test/functional/tests/io_class/test_io_classification.py [hunk 5]
```diff
@@ -566,6 +565,114 @@ def test_ioclass_conditions_and(filesystem):
         sync()
         new_occupancy = cache.get_statistics_deprecated(io_class_id=1)["occupancy"]
 
-        assert new_occupancy == base_occupancy, \
-            "Unexpected occupancy increase!\n" \
-            f"Expected: {base_occupancy}, actual: {new_occupancy}"
+        if new_occupancy != base_occupancy:
+            TestRun.fail("Unexpected occupancy increase!\n"
```
Comment (micrakow) [inline test/functional/tests/io_class/test_io_classification.py]: general - what if occupancy decreases? :dancing_women:

**Pull request merged**